src/overlords/offense/roomPoisoner.ts

```diff
@@ -0,0 +1,159 @@
+import {Zerg} from '../../zerg/Zerg';
+import {MY_USERNAME} from '../../~settings';
+import {Overlord} from '../Overlord';
+import {OvermindConsole} from '../../console/Console';
+import {ColonyMemory} from '../../Colony';
+
+/**
+ * Spawn roomPoisoner - upgrqde controller to lvl2, wall in controller then sources.
+ */
+@profile
+export class RoomPoisonerOverlord extends Overlord {
+
+	roomPoisoners: Zerg[];
+	antiControllers: Zerg[]; //claim and counter reserve if required
+
+    controllerWallSites: ConstructionSite[] | undefined;
+    sourcesWallSites: ConstructionSite[] | undefined;
+
+	constructor(directive: DirectivePoisonRoom, priority = OverlordPriority.offense.roomPoisoner) {
+		super(directive, 'contaminate', priority);
+		this.roomPoisoners = this.zerg(Roles.roomPoisoner);
+		this.antiControllers = this.zerg(Roles.claim);
+
+		this.controllerWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+                                                                    s => s.structureType == STRUCTURE_WALL &&
+                                                                    s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+    	this.sourcesWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+                                                                    s => s.structureType == STRUCTURE_WALL &&
+                                                                    !s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+    }
+
+	refresh() {
+		super.refresh();
+		this.controllerWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+            s => s.structureType == STRUCTURE_WALL &&
+            s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+        this.sourcesWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+            s => s.structureType == STRUCTURE_WALL &&
+            !s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+	}
+
+	init() {
+		let isSafe = this.room && !this.room.dangerousPlayerHostiles.length;
+		let isNotReservedByEnemy = !(this.room && this.room.controller && this.room.controller.reservation && this.room.controller.reservation.ticksToEnd > 500);
+
+		//spawn required creeps to contaminate if visible + safe + notRserved + notPoisoned, else spawn reserved is reserved by enemy
+		
+		if(this.pos.isVisible && isSafe && !this.initializer.memory.isPoisoned){
+			if(isNotReservedByEnemy){
+				if(!(this.room && this.room.my)) this.wishlist(1, Setups.infestors.claim);
+				this.wishlist(1, Setups.roomPoisoner);
+			}else{
+				if(!(this.room && this.room.my)) this.wishlist(1, Setups.infestors.controllerAttacker);	
+			}
+		}
+	}
+
+	private findStructureBlockingController(roomPoisoner: Zerg): Structure | undefined {
+		const blockingPos = Pathing.findBlockingPos(roomPoisoner.pos, roomPoisoner.room.controller!.pos,
```

Comment:
`Pathing.findBlockingPos` is expensive to call -- you should cache the result using `$.structures()`

Reply:
have just removed it, it is a corner case anyway.
but, findStructureBlockingController was actually taken from pioneer overlord for the same purpose and it was not cashed as described too.